datumaro/components/project.py

```diff
@@ -2,488 +2,2181 @@
 
-        env_dir = osp.join(project_dir, config.env_dir)
-        save_dir = osp.abspath(env_dir)
+            return self._RefKind.tree, obj_hash
+        except Exception as e:
+            if isinstance(e, AssertionError):
+                raise
 
-        project_dir_existed = osp.exists(project_dir)
-        env_dir_existed = osp.exists(env_dir)
         try:
-            os.makedirs(save_dir, exist_ok=True)
+            assert self._dvc.is_hash(ref), ref
+            return self._RefKind.blob, ref
+        except Exception as e:
+            raise UnknownRefError(ref) from e
+
+    def _materialize_rev(self, rev: Revision):
+        obj_dir = self.cache_path(rev)
+        if osp.isdir(obj_dir):
+            return obj_dir
+
+        tree = self._git.get_tree(rev)
+        self._git.write_tree(tree, obj_dir)
+        return obj_dir
+
+    def _is_cached(self, obj_hash: Reference):
+        return osp.isdir(self.cache_path(obj_hash))
+
+    def cache_path(self, obj_hash: Reference) -> str:
+        assert self._git.is_hash(obj_hash) or self._dvc.is_hash(obj_hash), obj_hash
+        if self._dvc.is_dir_hash(obj_hash):
+            obj_hash = obj_hash[:self._dvc.FILE_HASH_LEN]
+
+        return osp.join(self._aux_dir, ProjectLayout.cache_dir,
+            obj_hash[:2], obj_hash[2:])
+
+    def _can_retrieve_from_vcs_cache(self, obj_hash: Reference):
+        if not self._dvc.is_dir_hash(obj_hash):
+            dir_check = self._dvc.is_cached(
+                obj_hash + self._dvc.DIR_HASH_SUFFIX)
+        else:
+            dir_check = False
+        return dir_check or self._dvc.is_cached(obj_hash)
+
+    def source_data_dir(self, name: str) -> str:
+        return osp.join(self._root_dir, name)
+
+    def _source_dvcfile_path(self, name: str, root: str = None) -> str:
+        """
+        root - Path to the tree root directory. If not set,
+            the working tree is used.
+        """
+
+        if not root:
+            root = osp.join(self._aux_dir, ProjectLayout.tree_dir)
+        return osp.join(root, TreeLayout.sources_dir, name + '.dvc')
+
+    def _make_tmp_dir(self, suffix=None):
+        project_tmp_dir = osp.join(self._aux_dir, ProjectLayout.tmp_dir)
+        os.makedirs(project_tmp_dir, exist_ok=True)
+        if suffix:
+            suffix = '_' + suffix
+        return tempfile.TemporaryDirectory(suffix=suffix, dir=project_tmp_dir)
+
+    def remove_cache_obj(self, ref: Union[Revision, Reference]):
+        obj_type, obj_hash = self._parse_ref(ref)
+
+        if self._is_cached(obj_hash):
+            rmtree(self.cache_path(obj_hash))
+
+        if obj_type == self._RefKind.tree:
+            pass
+        elif obj_type == self._RefKind.blob:
+            self._dvc.remove_cache_obj(obj_hash)
+        else:
+            raise ValueError("Unexpected object type '%s'" % obj_type)
+
+    def validate_source_name(self, name: str):
+        disallowed_symbols = r"[^\\\ \.\~\-\_\w\d]"
```

Comment:
The regex could be written more readably as `r"[^\\ .~\w-]" (note that `\w` includes both digits and underscores).